File: src/application/commands/bootstrap_handler.py
"""BootstrapHandler -- application command for the bootstrap flow.

Orchestrates tool detection, preview generation, confirmation, and execution.
Depends on ports (abstractions), never on infrastructure directly.
"""

from __future__ import annotations

from pathlib import Path
from typing import TYPE_CHECKING

from src.domain.models.bootstrap_session import BootstrapSession, SessionNotFoundError
from src.domain.models.preview import FileAction, FileActionType, Preview

if TYPE_CHECKING:
    from src.application.ports.tool_detection_port import ToolDetectionPort

# Files alty plans to create in a new project.
_PLANNED_FILES: tuple[Path, ...] = (
    Path("docs/PRD.md"),
    Path("docs/DDD.md"),
    Path("docs/ARCHITECTURE.md"),
    Path("AGENTS.md"),
    Path(".alty/config.toml"),
    Path(".alty/knowledge/_index.toml"),
    Path(".alty/maintenance/doc-registry.toml"),
)


class BootstrapHandler:
    """Orchestrates the preview -> confirm -> execute bootstrap flow.

    Attributes:
        _tool_detection: Port for detecting installed AI coding tools.
        _sessions: In-memory store of active sessions (keyed by session_id).
    """

    def __init__(self, tool_detection: ToolDetectionPort) -> None:
        self._tool_detection = tool_detection
        self._sessions: dict[str, BootstrapSession] = {}

    def preview(self, project_dir: Path) -> BootstrapSession:
        """Create a new session and generate a preview of planned actions.

        Raises:
            FileNotFoundError: If the project directory has no README.md.
        """
        readme = project_dir / "README.md"
        if not readme.exists():
            raise FileNotFoundError("Create a README.md with your project idea first")

        detected_tools = self._tool_detection.detect(project_dir)
        conflict_descriptions = self._tool_detection.scan_conflicts(project_dir)

        file_actions: list[FileAction] = []
        for planned in _PLANNED_FILES:
            full_path = project_dir / planned
            if full_path.exists():
                file_actions.append(
                    FileAction(
                        path=planned,
                        action_type=FileActionType.SKIP,
                        reason="already exists",
                    )
                )
            else:
                file_actions.append(FileAction(path=planned, action_type=FileActionType.CREATE))

        preview = Preview(
            file_actions=tuple(file_actions),
            conflict_descriptions=tuple(conflict_descriptions),
        )
        session = BootstrapSession(project_dir=project_dir)
        session.set_detected_tools(detected_tools)
        session.set_preview(preview)
        self._sessions[session.session_id] = session
        return session

    def _get_session(self, session_id: str) -> BootstrapSession:
        """Look up a session by ID.

        Raises:
            SessionNotFoundError: If no session matches the given ID.
        """
        try:
            return self._sessions[session_id]
        except KeyError:
            raise SessionNotFoundError(f"No active session with id '{session_id}'") from None

    def confirm(self, session_id: str) -> BootstrapSession:
        """Confirm a previewed session, enabling execution."""
        session = self._get_session(session_id)
        session.confirm()
        return session

    def cancel(self, session_id: str) -> BootstrapSession:
        """Cancel a previewed session."""
        session = self._get_session(session_id)
        session.cancel()
        return session

    def execute(self, session_id: str) -> BootstrapSession:
        """Execute a confirmed session.

        Transitions through EXECUTING to COMPLETED. Emits BootstrapCompleted.
        """
        session = self._get_session(session_id)
        session.begin_execution()
        # Future: call DiscoveryPort, ArtifactRendererPort, etc.
        session.complete()
        return session

File: internal/bootstrap/application/bootstrap_handler.go
// Package application provides command handlers for the Bootstrap bounded context.
package application

import (
	"context"
	"fmt"
	"path/filepath"
	"sync"

	"github.com/alty-cli/alty/internal/bootstrap/domain"
	sharedapp "github.com/alty-cli/alty/internal/shared/application"
	vo "github.com/alty-cli/alty/internal/shared/domain/valueobjects"
)

// ToolDetector detects installed AI coding tools and scans for config conflicts.
// Defined where consumed per Go convention (no ctx parameter needed for sync ops).
type ToolDetector interface {
	Detect(projectDir string) ([]string, error)
	ScanConflicts(projectDir string) ([]string, error)
}

// FileChecker checks whether a file exists. Extracted as a port so the
// application layer has no direct os.Stat dependency.
type FileChecker interface {
	Exists(path string) bool
}

// ContentProvider returns generated content for a planned file path.
type ContentProvider interface {
	ContentFor(path string, config domain.ProjectConfig) string
}

// plannedFiles lists files alty plans to create in a new project.
var plannedFiles = []string{
	"docs/PRD.md",
	"docs/DDD.md",
	"docs/ARCHITECTURE.md",
	"AGENTS.md",
	".alty/config.toml",
	".alty/knowledge/_index.toml",
	".alty/maintenance/doc-registry.toml",
}

// ScaffoldCommitMessage is the commit message used when auto-committing scaffold files.
const ScaffoldCommitMessage = "chore: initialize alty project structure"

// BootstrapHandler orchestrates the preview -> confirm -> execute bootstrap flow.
type BootstrapHandler struct {
	toolDetection   ToolDetector
	fileChecker     FileChecker
	publisher       sharedapp.EventPublisher
	fileWriter      sharedapp.FileWriter
	contentProvider ContentProvider
	gitCommitter    GitCommitter
	mu              sync.Mutex
	sessions        map[string]*domain.BootstrapSession
	configs         map[string]domain.ProjectConfig
}

// BootstrapOption configures optional dependencies for BootstrapHandler.
type BootstrapOption func(*BootstrapHandler)

// WithGitCommitter injects an optional GitCommitter for auto-committing scaffold files.
func WithGitCommitter(gc GitCommitter) BootstrapOption {
	return func(h *BootstrapHandler) {
		h.gitCommitter = gc
	}
}

// NewBootstrapHandler creates a new BootstrapHandler with injected dependencies.
func NewBootstrapHandler(toolDetection ToolDetector, fileChecker FileChecker, publisher sharedapp.EventPublisher, fileWriter sharedapp.FileWriter, contentProvider ContentProvider, opts ...BootstrapOption) *BootstrapHandler {
	h := &BootstrapHandler{
		toolDetection:   toolDetection,
		fileChecker:     fileChecker,
		publisher:       publisher,
		fileWriter:      fileWriter,
		contentProvider: contentProvider,
		sessions:        make(map[string]*domain.BootstrapSession),
		configs:         make(map[string]domain.ProjectConfig),
	}
	for _, opt := range opts {
		opt(h)
	}
	return h
}

// Preview creates a new session and generates a preview of planned actions.
func (h *BootstrapHandler) Preview(projectDir string) (*domain.BootstrapSession, error) {
	readme := filepath.Join(projectDir, "README.md")
	if !h.fileChecker.Exists(readme) {
		return nil, fmt.Errorf("create a README.md with your project idea first")
	}

	tools, err := h.toolDetection.Detect(projectDir)
	if err != nil {
		return nil, fmt.Errorf("detecting tools: %w", err)
	}

	conflicts, err := h.toolDetection.ScanConflicts(projectDir)
	if err != nil {
		return nil, fmt.Errorf("scanning conflicts: %w", err)
	}

	var fileActions []vo.FileAction
	for _, planned := range plannedFiles {
		fullPath := filepath.Join(projectDir, planned)
		if h.fileChecker.Exists(fullPath) {
			fileActions = append(fileActions,
				vo.NewFileAction(planned, vo.FileActionSkip, "already exists", ""))
		} else {
			fileActions = append(fileActions,
				vo.NewFileAction(planned, vo.FileActionCreate, "", ""))
		}
	}

	preview := vo.NewPreview(fileActions, nil, conflicts)
	session := domain.NewBootstrapSession(projectDir)
	session.SetDetectedTools(tools)
	if err := session.SetPreview(&preview); err != nil {
		return nil, fmt.Errorf("setting preview: %w", err)
	}

	h.mu.Lock()
	h.sessions[session.SessionID()] = session
	h.mu.Unlock()
	return session, nil
}

// SetGitCommitter sets or clears the GitCommitter at runtime.
// Used by the CLI to honor the --no-commit flag.
func (h *BootstrapHandler) SetGitCommitter(gc GitCommitter) {
	h.gitCommitter = gc
}

// WithProjectConfig associates a ProjectConfig with a session. Must be called
// before Execute so that content generation receives detected project settings.
func (h *BootstrapHandler) WithProjectConfig(sessionID string, config domain.ProjectConfig) {
	h.mu.Lock()
	h.configs[sessionID] = config
	h.mu.Unlock()
}

// Confirm confirms a previewed session, enabling execution.
func (h *BootstrapHandler) Confirm(sessionID string) (*domain.BootstrapSession, error) {
	session, err := h.getSession(sessionID)
	if err != nil {
		return nil, err
	}
	if err := session.Confirm(); err != nil {
		return nil, fmt.Errorf("confirm session: %w", err)
	}
	return session, nil
}

// Cancel cancels a previewed session.
func (h *BootstrapHandler) Cancel(sessionID string) (*domain.BootstrapSession, error) {
	session, err := h.getSession(sessionID)
	if err != nil {
		return nil, err
	}
	if err := session.Cancel(); err != nil {
		return nil, fmt.Errorf("cancel session: %w", err)
	}
	return session, nil
}

// Execute executes a confirmed session.
func (h *BootstrapHandler) Execute(sessionID string) (*domain.BootstrapSession, error) {
	session, err := h.getSession(sessionID)
	if err != nil {
		return nil, err
	}
	if err := session.BeginExecution(); err != nil {
		return nil, fmt.Errorf("begin execution: %w", err)
	}

	h.mu.Lock()
	config, hasConfig := h.configs[sessionID]
	h.mu.Unlock()

	if !hasConfig {
		// Fallback: build minimal config from session data.
		config = domain.NewProjectConfig(
			filepath.Base(session.ProjectDir()),
			"", "", session.DetectedTools(),
		)
	}

	var writtenPaths []string
	preview := session.Preview()
	if preview != nil {
		for _, action := range preview.FileActions() {
			if action.ActionType() != vo.FileActionCreate {
				continue
			}
			content := h.contentProvider.ContentFor(action.Path(), config)
			target := filepath.Join(session.ProjectDir(), action.Path())
			if err := h.fileWriter.WriteFile(context.Background(), target, content); err != nil {
				return nil, fmt.Errorf("writing %s: %w", action.Path(), err)
			}
			writtenPaths = append(writtenPaths, action.Path())
		}
	}

	if err := h.commitScaffold(context.Background(), session.ProjectDir(), writtenPaths); err != nil {
		return nil, fmt.Errorf("committing scaffold: %w", err)
	}

	if err := session.Complete(); err != nil {
		return nil, fmt.Errorf("complete session: %w", err)
	}
	for _, event := range session.Events() {
		_ = h.publisher.Publish(context.Background(), event)
	}
	return session, nil
}

func (h *BootstrapHandler) getSession(sessionID string) (*domain.BootstrapSession, error) {
	h.mu.Lock()
	session, ok := h.sessions[sessionID]
	h.mu.Unlock()
	if !ok {
		return nil, fmt.Errorf("no active session with id '%s'", sessionID)
	}
	return session, nil
}

// commitScaffold stages and commits written files if a GitCommitter is configured.
// Skips gracefully when: no GitCommitter, not a git repo, or no files to commit.
func (h *BootstrapHandler) commitScaffold(ctx context.Context, projectDir string, paths []string) error {
	if h.gitCommitter == nil || len(paths) == 0 {
		return nil
	}

	hasGit, err := h.gitCommitter.HasGit(ctx, projectDir)
	if err != nil {
		return fmt.Errorf("checking git repo: %w", err)
	}
	if !hasGit {
		return nil
	}

	if err := h.gitCommitter.StageFiles(ctx, projectDir, paths); err != nil {
		return fmt.Errorf("staging files: %w", err)
	}

	if err := h.gitCommitter.Commit(ctx, projectDir, ScaffoldCommitMessage); err != nil {
		return fmt.Errorf("creating commit: %w", err)
	}

	return nil
}
